src/ReductionMsg.chpl

```diff
@@ -611,11 +619,39 @@ module ReductionMsg
       return res;
     }
     
-    proc segMean(values:[] ?t, segments:[?D] int): [D] real {
+    proc segMean(values:[] ?t, segments:[?D] int, skipNan=false): [D] real {
       var res: [D] real;
       if (D.size == 0) { return res; }
-      var sums = segSum(values, segments);
-      var counts = segCount(segments, values.size);
+      var sums;
+      var counts;
+      if (isFloatType(t) && skipNan) {
+        // count cumulative nans over all values
+        var cumnans = isnan(values):int;
+        cumnans = + scan cumnans;
+        
+        // find cumulative nans at segment boundaries
+        var segnans: [segments.domain] int;
+        forall si in segments.domain {
+          if si == segments.domain.high {
+              segnans[si] = cumnans[cumnans.domain.high];
+          } else {
+              segnans[si] = cumnans[segments[si+1]-1];
+          }
+        }
+        
+        // take diffs of adjacent segments to find nan count in each segment
+        var nancounts: [segments.domain] int;
+        nancounts[0] = segnans[0];
```

Comment:
In case the default domain ever changes, we should probably use implicit indexing here, i.e. `nancounts[segments.domain.low] = segnans[segments.domain.low];`